SU2_CFD/include/gradients/computeGradientsLeastSquares.hpp

```diff
@@ -152,155 +146,190 @@ void computeGradientsLeastSquares(CSolver* solver,
       }
     }
 
-    for (size_t iDim = 0; iDim < nDim; ++iDim)
-      for (size_t jDim = 0; jDim < nDim; ++jDim)
-        AD::SetPreaccOut(Rmatrix(iPoint, iDim, jDim));
+    if (periodic)
+    {
+      /*--- A second loop is required after periodic comms, checkpoint the preacc. ---*/
 
-    for (size_t iVar = varBegin; iVar < varEnd; ++iVar)
       for (size_t iDim = 0; iDim < nDim; ++iDim)
-        AD::SetPreaccOut(gradient(iPoint, iVar, iDim));
+        for (size_t jDim = 0; jDim < nDim; ++jDim)
+          AD::SetPreaccOut(Rmatrix(iPoint, iDim, jDim));
 
-    AD::EndPreacc();
+      for (size_t iVar = varBegin; iVar < varEnd; ++iVar)
+        for (size_t iDim = 0; iDim < nDim; ++iDim)
+          AD::SetPreaccOut(gradient(iPoint, iVar, iDim));
+
+      AD::EndPreacc();
+    }
+    else {
+      /*--- Periodic comms are not needed, solve the LS problem for iPoint. ---*/
+
+      solveLeastSquares(iPoint, false, nDim, varBegin, varEnd, Rmatrix, gradient);
+    }
   }
 
   /*--- Correct the gradient values across any periodic boundaries. ---*/
 
-  if (solver != nullptr)
+  if (periodic)
   {
     for (size_t iPeriodic = 1; iPeriodic <= config.GetnMarker_Periodic()/2; ++iPeriodic)
     {
       solver->InitiatePeriodicComms(&geometry, &config, iPeriodic, kindPeriodicComm);
       solver->CompletePeriodicComms(&geometry, &config, iPeriodic, kindPeriodicComm);
     }
+
+    /*--- Second loop over points of the grid to compute final gradient. ---*/
+
+    SU2_OMP_FOR_DYN(chunkSize)
+    for (size_t iPoint = 0; iPoint < nPointDomain; ++iPoint)
+      solveLeastSquares(iPoint, true, nDim, varBegin, varEnd, Rmatrix, gradient);
   }
 
-  /*--- Second loop over points of the grid to compute final gradient. ---*/
+  /*--- If no solver was provided we do not communicate ---*/
 
-  SU2_OMP_FOR_DYN(chunkSize)
-  for (size_t iPoint = 0; iPoint < nPointDomain; ++iPoint)
+  if (solver != nullptr)
   {
-    /*--- Entries of upper triangular matrix R. ---*/
+    /*--- Obtain the gradients at halo points from the MPI ranks that own them. ---*/
+
+    solver->InitiateComms(&geometry, &config, kindMpiComm);
+    solver->CompleteComms(&geometry, &config, kindMpiComm);
+  }
 
-    su2double r11 = Rmatrix(iPoint,0,0);
-    su2double r12 = Rmatrix(iPoint,0,1);
-    su2double r22 = Rmatrix(iPoint,1,1);
-    su2double r13 = 0.0, r23 = 0.0, r23_a = 0.0, r23_b = 0.0, r33 = 0.0;
+}
+
+/*!
+ * \brief Solve the LS problem prepared above.
+ */
+template<class GradientType, class RMatrixType>
+FORCEINLINE void solveLeastSquares(size_t iPoint,
+                                   bool periodic,
+                                   size_t nDim,
+                                   size_t varBegin,
+                                   size_t varEnd,
+                                   const RMatrixType& Rmatrix,
+                                   GradientType& gradient) {
+  constexpr size_t MAXNDIM = 3;
```

Comment:
This seems like something we should have defined in one place, like option_structure.hpp